execution_engine/src/runtime/externals.rs

```diff
@@ -1184,6 +1184,9 @@ where
                 let message = self.t_from_mem(message_ptr, message_size)?;
 
                 let result = self.emit_message(topic_name, message)?;
+                if result.is_ok() {
+                    self.charge_emit_message()?;
```

Comment:
Are these calls immediate? I.e. if we fail to charge, do we still emit the message?

Reply:
The effects should be reverted if we fail to charge since we would trap with a `GasLimit` error.
What I wanted to do is charge if the message was actually emitted. If an `ApiError` occurred, we shouldn't charge for it. Are you suggesting we should charge every attempt to emit a message? Then the next attempt would have increased gas or have the previous gas cost?

Reply:
As discussed on slack we should always charge, but increase the counter if the message is actually emitted - in this case, the user has a chance to retry the message by (for instance) shrinking the message to fit the limit.